### ISSUE TITLE
How do I perform a calculations (division) in a report column

### ISSUE THREAD

**Issue by lie112** [APSIMInitiative/ApsimX#5661]
I haven't been able to find an example of dividing two properties in a report column and this isn't covered in the report documentation.
I found the Divide() Expression. I assume I pass the parameters as a comma delimited list.
In ExpressionEvaluator.EvaluatePostFix() the tpStack has both properties passed, but tpStack.Pop() only returns the second and so the expression fails and produces the generic error "Error in report XXXX: Invalid report variables found: XXXX
Reporting errors could be improved to provide a better explanation of what went wrong.

I was trying 
Divide([Activities].ReportHerd.ReportDetails.Individual.Intake, [Activities].ReportHerd.ReportDetails.Individual.PotentialIntake) as PropIntake

Can I also use two previously created columns (e.g. xxxx as A and yyyy as B) in this equation or only properties that can be located? i.e.  Divide(A,B) as C

**Comment by hol430** [comment]
For dividing 2 properties - you should be able to use the `/` operator - is this not working for you? I tried this and it worked for me:

```
[Clock].Today.DayOfYear / [Clock].Today.Day as test
```

For the second part - yes this would be a useful feature, I guess we would just need some way to hijack the mechanism by which variable names are translated into variable values (e.g. as in `ExpressionFunction.FillVariableNames()`). I haven't really looked into the expression parser code recently, so I can't really make a more specific comment on this one.

**Comment by lie112** [comment]
Your example works in my report and I thought that was the case.

I have just tried:

`[Activities].ReportHerd.ReportDetails.Individual.Intake as Intake2
[Activities].ReportHerd.ReportDetails.Individual.PotentialIntake as PotentialIntake2
[Activities].ReportHerd.ReportDetails.Individual.Intake / [Activities].ReportHerd.ReportDetails.Individual.PotentialIntake`

and got
`Error in report HerdReport: Invalid report variables found:
Activities.ReportHerd.ReportDetails.Individual.Intake / Activities.ReportHerd.ReportDetails.Individual.PotentialIntake
`

The previous two lines work, so the variables are correct and present.
Intake and Potential are both public double properties { get; set; } of the Ruminant (Individual) and show up in intellisense
+ * and ^ all work, but not / for my case

**Comment by hol430** [comment]
Could it be a divide by 0 problem? Otherwise I'm not sure what the problem is - it seems strange that the other operators work. If you upload your .apsimx file I can look into it.

**Comment by lie112** [comment]
[TestReportCalculation.zip](https://github.com/APSIMInitiative/ApsimX/files/5221962/TestReportCalculation.zip)
Thanks @hol430 
I didn't spend anymore time searching for the cause as I wasn't sure I was even doing division properly.

**Comment by hol430** [comment]
Ok yep it's a division by zero problem. I've tweaked the error reporting code in the report component, so I now get the following message in the status panel:

```
Error in report HerdReport: Invalid report variables found:
IntakeRatio: Divide by Zero.

----------------------------------------------
```

Which should help when diagnosing similar problems in the future, but is probably less than helpful for you right now. I'm not sure what the best solution is to this problem...what did old apsim do in this scenario?

**Comment by lie112** [comment]
Thanks. Better error reporting in Reports will be useful.
Now I need to work out why there are entries with 0 potential intake.

**Comment by lie112** [comment]
Resolved #5667

**Comment by hol430** [comment]
Ok I'll close this then. Feel free to reopen if need be.
